test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/TestFhirServer.cs

```diff
@@ -52,26 +69,125 @@ public TestFhirClient GetTestFhirClient(ResourceFormat format, TestApplication c
 
             if (!reusable)
             {
-                return CreateFhirClient(format, clientApplication, user);
+                return CreateFhirClient(format, clientApplication, user, authenticationHandler);
             }
 
             return _cache.GetOrAdd(
                     (format, clientApplication, user),
                     (tuple, fhirServer) =>
-                        new Lazy<TestFhirClient>(() => CreateFhirClient(tuple.format, tuple.clientApplication, tuple.user)),
+                        new Lazy<TestFhirClient>(() => CreateFhirClient(tuple.format, tuple.clientApplication, tuple.user, authenticationHandler)),
                     this)
                 .Value;
         }
 
-        private TestFhirClient CreateFhirClient(ResourceFormat format, TestApplication clientApplication, TestUser user)
+        private TestFhirClient CreateFhirClient(ResourceFormat format, TestApplication clientApplication, TestUser user, DelegatingHandler authenticationHandler = null)
         {
-            var httpClient = new HttpClient(new SessionMessageHandler(CreateMessageHandler(), _asyncLocalSessionTokenContainer)) { BaseAddress = BaseAddress };
+            ConfigureSecurityOptions().GetAwaiter().GetResult();
```

Comment:
I wonder if this could just be part of the SendAsync in chained the handler?

Something like FhirClient -> HttpClient -> AuthenticationHandler -> SessionMessageHandler

AuthenticationHandler.SendAsync will do a few things. 1. Read CapatibilityStatement if it hasn't. 2. If SecurityOption is enabled, then delegate to some external class to configure and create appropriate handler/provider that can get/set token. 3. Cache the token and check expiration date and starts refresh if it's within X minutes.

Reply:
Realistically, this only needs to be executed once per server. I'm going to look for a way to move this out of client creation and move it into when the `TestFhirServer` is realized. 

I don't think the `AuthenticationHandler` should be concerned with this because this is really a test specific problem in my opinion. In normal uses of the `Client` I would expect the developer to know how they're going to obtain a token and there would be no need to dynamically do it based on configuration exposed by the server.